core-backend/common/src/lib.rs

```diff
@@ -194,12 +194,12 @@ pub trait Environment<E: Ext + IntoExtInfo + 'static>: Sized {
     /// 2) Creates wasm memory
     /// 3) Runs `pre_execution_handler` to fill the memory before running instance.
     /// 4) Instantiate external funcs for wasm module.
-    /// 5) Run instance setup starting at `entry_point` - wasm export function name.
+    fn new(ext: E, binary: &[u8], mem_size: WasmPageNumber) -> Result<Self, Self::Error>;
+
+    /// Run instance setup starting at `entry_point` - wasm export function name.
     fn execute<F, T>(
-        ext: E,
-        binary: &[u8],
+        self,
```

Comment:
```suggestion
        &self,
```

Can't we use self by ref for future purposes?

Reply:
Such changes require a lot of refactoring so it must be separate PR